SpatialGDKEditor/Source/Private/SpatialGDKEditor.cpp

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialGDKEditor.h"
+
+#include "Async.h"
+#include "SpatialGDKEditorSchemaGenerator.h"
+
+#include "Editor.h"
+
+#include "AssetRegistryModule.h"
+#include "GeneralProjectSettings.h"
+
+DEFINE_LOG_CATEGORY(LogSpatialGDKEditor);
+
+USpatialGDKEditor::USpatialGDKEditor()
+	: bSchemaGeneratorRunning(false)
+{
+
+}
+
+void USpatialGDKEditor::GenerateSchema(FSimpleDelegate SuccessCallback, FSimpleDelegate FailureCallback, FSpatialGDKEditorErrorHandler ErrorCallback)
+{
+	if (bSchemaGeneratorRunning)
+	{
+		UE_LOG(LogSpatialGDKEditor, Warning, TEXT("Schema generation is already running"));
+		return;
+	}
+
+	bSchemaGeneratorRunning = true;
+
+	// Force spatial networking so schema layouts are correct
+	UGeneralProjectSettings* GeneralProjectSettings = GetMutableDefault<UGeneralProjectSettings>();
+	bool bCachedSpatialNetworking = GeneralProjectSettings->bSpatialNetworking;
+	GeneralProjectSettings->bSpatialNetworking = true;
+
+
+	// Ensure all our spatial classes are loaded into memory before running
+	CacheSpatialObjects(SPATIALCLASS_GenerateTypeBindings, ErrorCallback);
+
+	SchemaGeneratorResult = Async<bool>(EAsyncExecution::Thread, SpatialGDKGenerateSchema,
+		[this, bCachedSpatialNetworking, SuccessCallback, FailureCallback]()
+	{
+		if (!SchemaGeneratorResult.IsReady() || SchemaGeneratorResult.Get() != true)
+		{
+			FailureCallback.Execute();
+			bSchemaGeneratorRunning = false;
+		}
+		else
+		{
+			SuccessCallback.Execute();
+			bSchemaGeneratorRunning = false;
+		}
+		GetMutableDefault<UGeneralProjectSettings>()->bSpatialNetworking = bCachedSpatialNetworking;
+		bSchemaGeneratorRunning = false;
+	});
+}
+
+void USpatialGDKEditor::GenerateSnapshot(UWorld* World, FString SnapshotFilename, FSimpleDelegate SuccessCallback, FSimpleDelegate FailureCallback, FSpatialGDKEditorErrorHandler ErrorCallback)
+{
+	// Ensure all our singletons are loaded into memory before running
+	CacheSpatialObjects(SPATIALCLASS_Singleton, ErrorCallback);
+
+	const bool bSuccess = SpatialGDKGenerateSnapshot(World, SnapshotFilename);
+
+	if (bSuccess)
+	{
+		SuccessCallback.Execute();
```

Comment:
Same question as in `GenerateSchema` - check `IfBound` unless it's invalid to call with empty callback.